### ISSUE TITLE
Problem training Yolo on handwritten digits

### ISSUE THREAD

**Issue by Astrus1234** [pjreddie/darknet#643]
Hello,

I have a dataset of handwritten digits that look like this : https://pasteboard.co/HfaR853.jpg
My goal is to be able to separate each digits in order to feed this new data into a neural network that is able to predict which digit it is (0-9 range).

I had the idea to try out YOLO for this problem but it's not working at all. Here is what I do : 

I have 100 images with 10 digits on each (1000 digits) with the corresponding bounding box that I annoted manually. I choosed to annoted 0-9 numbers under the same category "NUMBER" but I dont know if it's the correct way ? 

I tried to overfit Yolo on 4 images to see if it's appropriate for my problem as suggested by others members.

Here is what I obtain after long time of training : https://pasteboard.co/HfaTGq0.png

The result that I obtain on one of the same image use as training (should overfit) : https://pasteboard.co/HfaTeUI.jpg

You can see that I obtain a good IOU but it's not working at all... I dont know if I'm not doing something correctly in the config file ? I have changed the output according to the rule :  (1+5)*5 = 30

Thank's for the help,
Astrus

**Comment by TheMikeyR** [comment]
The image is 506x68, what is the size your network? It is recommended that the images you input is at least at the minimum network size. You could e.g. pad the image to make it bigger and try to retrain.

**Comment by TheMikeyR** [comment]
Also according to the config file it is (classes + coords + 1) * num, so if you have 10 numbers (0-9) it should be (10 + 4 + 1) * 5 = 75, but if you only tried to train with one class then 30 is correct.

**Comment by TheMikeyR** [comment]
It is recommended to keep the network in square so e.g. 416x416, Didn't hear about anyone who managed to train it well on non-square network, though you can still try it, but as you have problems with training you should start with something that is supposed to work and then you can try non-square when you got it to work with a square network. 

I don't know if it's enough with only 4 images, I'm not experienced with low amount of images. 

About every number (0-9) your class is "number" should work, but it might be easier/quicker for the network to train with a classname for each number. But it should definitely be possible to train with just "number" for all numbers. So it's all up to you.

**Comment by Astrus1234** [comment]
I have used a 512x512 network with padding added to the images. I'm using 5 images to try to overfit them as much as possible. If it's proven to be working I'll train on the full dataset.

I'm actually at iteration 700 (5 images per iterations) and get the following result : 
https://pasteboard.co/HfcJZYe.png

I'll try to let it run for 2000 iterations at least.

I have multiple questions : 
Do you think the result can still be improved ?
How much iterations is necessary in general ?
Do you know how can I get the corresponding bounding boxes by coordinate in the image space ?

**Comment by TheMikeyR** [comment]
**Do you think the result can still be improved ?**
Yes, much bigger dataset (a rule of thumb is 2000 images per class), image augmentation etc. etc. many tricks are available and it's more or less trial and error. Also `random=1` in the cfg file usually give an increase of 1% mAP. 

**How much iterations is necessary in general ?**
Depending on the repository you are using, in AlexeyAB's fork he mention 2000 iterations per class. 

**Do you know how can I get the corresponding bounding boxes by coordinate in the image space ?**
I don't understand the question, if you want to get the coordinates out from the detections I'm using the python wrapper, you can get boxes by doing this https://github.com/pjreddie/darknet/issues/289#issuecomment-342448358

**Do you know a way to save the weight every 100 iterations ? (after 1000 it's saving only after another 1000 steps)**
Go into this file and line https://github.com/pjreddie/darknet/blob/76a0d8fb6d4faf4179d7d12ee6e5007a7408b643/examples/detector.c#L138 and change 10000 to 100.

**Comment by Astrus1234** [comment]
I have trained for the entire night 6000 iterations (5 images per iteration with 10 digits in each images).
I have obtain the following : https://pasteboard.co/HfkzNjL.png

When you mean 2000 iterations per class do you mean a full pass on each images ? I have 89 images in total that would mean 89/5 * 2000 = 35600 iterations ?

My question about the bounding boxes was about getting the coordinates out from the detection, so your method is working :+1: 

Thanks to your help the detection is getting better, thanks you so much :)

Here is the result on a trained image (should overfit if it's working) : https://pasteboard.co/HfkBG7q.jpg
Here is the result on a unseen image : https://pasteboard.co/HfkBYzX.jpg

Correct me if I'm wrong : the detection seems to works pretty well but it seems that the model could be trained a lot more 

Thank you again for all the help TheMikeyR  :+1:

**Comment by TheMikeyR** [comment]
You should increase your batch size to be 64, if you have trouble with generalising the model, since it will help generalize better, and then you probably need to increase subdivision if you can't hold it all in the gpu memory (if you get errors) e.g. 8. 

2000 iterations per class, if you have one class called "number" you should train 2000 steps (if batch size is 5 images, then you will have trained on 10000 images) at the 2000 iteration. If you have 10 classes (0-9) then it would need 20000 iterations. 
This is for AlexeyAB's fork, I've witnessed that original darknet require more iterations on my dataset than Alexey's fork. 

But looks like the last two result are much better, congratulations with that :-) Now it will be a bit of trial and error to see if you can improve it even more. A general tip is to get much more data than only 89 images, as said 2000 images per class. So if you are going for the "number" detection approach, you should get 2000 images of those you are training with. You could achieve this by augmenting, adding salt&pepper noise etc. etc.

**Comment by AhmedKhaled945** [comment]
Hello, i was wondering if i can access this dataset for research means, this would be wonderful, 
Thanks in advance. @Astrus1234
